go/ql/lib/semmle/go/frameworks/Fasthttp.qll

```diff
@@ -0,0 +1,372 @@
+/**
+ * Provides classes for working with untrusted flow sources, sinks and taint propagators
+ * from the `github.com/valyala/fasthttp` package.
+ */
+
+import go
+private import semmle.go.security.RequestForgeryCustomizations
+
+/**
+ * Provides classes for working with the [fasthttp](github.com/valyala/fasthttp) package.
+ */
+module Fasthttp {
+  /** Gets the v1 module path `github.com/valyala/fasthttp`. */
+  string v1modulePath() { result = "github.com/valyala/fasthttp" }
+
+  /** Gets the path for the root package of fasthttp. */
+  string packagePath() { result = package(v1modulePath(), "") }
+
+  /**
+   * Provide models for sanitizer/Dangerous Functions of fasthttp
+   */
+  module Functions {
+    /**
+     * A function that doesn't sanitize user-provided file paths
+     */
+    class FileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
+      FileSystemAccess() {
+        exists(Function f |
+          f.hasQualifiedName(packagePath(),
+            [
+              "ServeFile", "ServeFileUncompressed", "ServeFileBytes", "ServeFileBytesUncompressed",
+              "SaveMultipartFile"
+            ]) and
+          this = f.getACall()
+        )
+      }
+
+      override DataFlow::Node getAPathArgument() { result = this.getArgument(1) }
+    }
+
+    /**
+     * A function that can be used as a sanitizer for XSS
+     */
+    class HtmlQuoteSanitizer extends SharedXss::Sanitizer {
+      HtmlQuoteSanitizer() {
+        exists(DataFlow::CallNode c |
+          c.getTarget()
+              .hasQualifiedName(packagePath(),
+                ["AppendHTMLEscape", "AppendHTMLEscapeBytes", "AppendQuotedArg"])
```

Comment:
I've looked into this some more. I see now that no other libraries extend `SharedXss::Sink`. What they actually do is extend `Http::ResponseBody::Range`, like [here](https://github.com/github/codeql/blob/68a7734e08aff948de63c67c6d17c1195f2e968e/go/ql/lib/semmle/go/frameworks/Beego.qll#L171-L203). This has the same effect, because of [this class](https://github.com/github/codeql/blob/68a7734e08aff948de63c67c6d17c1195f2e968e/go/ql/lib/semmle/go/security/Xss.qll#L45-L47), but it also means you have to specify the content type. Please can you adjust your Shared XSS sinks to work this way.

I also found [this existing sanitizer](https://github.com/github/codeql/blob/68a7734e08aff948de63c67c6d17c1195f2e968e/go/ql/lib/semmle/go/security/Xss.qll#L117-L134), which makes me think your sanitizers are better modelled as extending [HtmlEscapeFunction](https://github.com/github/codeql/blob/68a7734e08aff948de63c67c6d17c1195f2e968e/go/ql/lib/semmle/go/Concepts.qll#L151-L158) or [UrlEscapeFunction](https://github.com/github/codeql/blob/68a7734e08aff948de63c67c6d17c1195f2e968e/go/ql/lib/semmle/go/Concepts.qll#L160-L166) as appropriate.

Reply:
For the first part: I don't think you should be trying to model this all yourself. We already have modelling for this, in `go/ql/lib/semmle/go/concepts/HTTP.qll`. Maybe you could skim that file, read the bits that seem relevant, and look at how they are used by models for other libraries. Or another approach would be to pick a library that you know and see how it is modelled. I'm not particularly familiar with our modelling in this area, but it looks like you want to write a class extending `Http::HeaderWrite::Range` to model all the ways that there are to write to headers in this library. And I believe you'll have to write a class extending `Http::ResponseBody::Range`, as I mentioned elsewhere. And I think `Http::ResponseWriter::Range`. If you look [here](https://github.com/github/codeql/blob/9fa20f5f39c1a15d288b4655edd2f07bac8de726/go/ql/lib/semmle/go/concepts/HTTP.qll#L181-L196) you'll see the logic for working out what the content type of a response body is.

Reply:
In fact, a really good approach would be to read the model for Beego (for example) to see what needs to be modelled for it, and model the same for FastHttp.